File: packages/sdk/src/runner.ts
/**
 * Runner API — mint, list, and revoke runner tokens.
 *
 * Maps to `/api/v1/private/runner/tokens`. Runner tokens (`mel_run_` prefix)
 * authenticate the `@melaya/runner` CLI process that executes agent pipelines
 * on your infrastructure.
 *
 * @example
 * ```ts
 * const { token } = await melaya.runner.createToken({ label: "prod-server-1" });
 * // Store token securely — it is shown only once.
 *
 * const tokens = await melaya.runner.listTokens();
 * await melaya.runner.revokeToken(tokens[0].id);
 * ```
 */
import type { HttpClient } from "./client.js";
import type { RunnerToken, RunnerTokenCreate, RunnerTokenCreateResult } from "./platform-types.js";

export class RunnerAPI {
  constructor(private readonly http: HttpClient) {}

  /**
   * Mint a new runner token.
   * The plaintext token is returned only in this response — store it securely.
   */
  async createToken(body?: RunnerTokenCreate): Promise<RunnerTokenCreateResult> {
    return this.http.post<RunnerTokenCreateResult>("/api/v1/private/runner/tokens", body);
  }

  /** List all runner tokens for the caller (masked, with last_seen). */
  async listTokens(): Promise<RunnerToken[]> {
    return this.http.get<RunnerToken[]>("/api/v1/private/runner/tokens");
  }

  /** Revoke a runner token by ID. */
  async revokeToken(tokenId: string): Promise<{ ok: boolean }> {
    return this.http.delete<{ ok: boolean }>(
      `/api/v1/private/runner/tokens/${encodeURIComponent(tokenId)}`,
    );
  }
}

File: packages/sdk-go/melaya/runner.go
// Runner API — mint, list, and revoke runner tokens (mel_run_ prefix).
//
// Maps to /api/v1/private/runner/tokens. Runner tokens authenticate the
// Melaya runner process that executes agent pipelines on your infrastructure.
//
// Example:
//
//	result, _ := m.Runner.CreateToken(ctx, &melaya.RunnerTokenCreate{Label: "prod-1"})
//	// store result.Token securely — it is shown only once
//	tokens, _ := m.Runner.ListTokens(ctx)
//	m.Runner.RevokeToken(ctx, tokens[0].ID)
package melaya

import (
	"context"
	"net/url"
)

// RunnerAPI wraps runner token endpoints.
type RunnerAPI struct {
	h *httpClient
}

// CreateToken mints a new runner token.
// The plaintext token is returned only in this response — store it securely.
//
// POST /api/v1/private/runner/tokens
func (r *RunnerAPI) CreateToken(ctx context.Context, body *RunnerTokenCreate) (*RunnerTokenCreateResult, error) {
	data, err := r.h.post(ctx, "/api/v1/private/runner/tokens", body)
	if err != nil {
		return nil, err
	}
	var v RunnerTokenCreateResult
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return &v, nil
}

// ListTokens lists all runner tokens for the caller (masked, with last_seen).
//
// GET /api/v1/private/runner/tokens
func (r *RunnerAPI) ListTokens(ctx context.Context) ([]RunnerToken, error) {
	data, err := r.h.get(ctx, "/api/v1/private/runner/tokens", nil)
	if err != nil {
		return nil, err
	}
	var v []RunnerToken
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return v, nil
}

// RevokeToken revokes a runner token by ID.
//
// DELETE /api/v1/private/runner/tokens/:tokenId
func (r *RunnerAPI) RevokeToken(ctx context.Context, tokenID string) (bool, error) {
	path := "/api/v1/private/runner/tokens/" + url.PathEscape(tokenID)
	data, err := r.h.del(ctx, path, nil)
	if err != nil {
		return false, err
	}
	var v okResult
	if err := unmarshal(data, &v); err != nil {
		return false, err
	}
	return v.Ok, nil
}
